runtime/marshaler_registry.go

```diff
@@ -5,132 +5,87 @@ import (
 	if outbound == nil {
-		outbound = defaultMarshaler
+		outbound = mux.marshalers.mimeMap[MIMEWildcard]
 	}
-	return inbound, outbound
 
+	return inbound, outbound
 }
 
-// marshalerRegistry keeps a mapping from MIME types to mimeMarshalers.
-type marshalerRegistry map[string]*mimeMarshaler
-
-type mimeMarshaler struct {
-	inbound  Marshaler
-	outbound Marshaler
+// marshalerRegistry is a mapping from MIME types to Marshalers.
+type marshalerRegistry struct {
+	mimeMap map[string]Marshaler
 }
 
-// addMarshaler adds an inbound and outbund marshaler for a case-sensitive MIME type string ("*" to match any MIME type).
-// Inbound is the marshaler that is used when marshaling inbound requests from the client.
-// Outbound is the marshaler that is used when marshaling outbound responses to the client.
-func (r *marshalerRegistry) add(mime string, inbound, outbound Marshaler) error {
-	if mime == "" {
+// add adds a marshaler for a case-sensitive MIME type string ("*" to match any
+// MIME type).
+func (m marshalerRegistry) add(mime string, marshaler Marshaler) error {
+	if len(mime) == 0 {
 		return errors.New("empty MIME type")
 	}
-	(*r)[mime] = &mimeMarshaler{
-		inbound:  inbound,
-		outbound: outbound,
-	}
-	return nil
-}
 
-// addInboundMarshaler adds an inbound marshaler for a case-sensitive MIME type string ("*" to match any MIME type).
-// Inbound is the marshaler that is used when marshaling inbound requests from the client.
-func (r *marshalerRegistry) addInbound(mime string, inbound Marshaler) error {
-	if mime == "" {
-		return errors.New("empty MIME type")
-	}
-	if entry := (*r)[mime]; entry != nil {
-		entry.inbound = inbound
-		return nil
-	}
-	(*r)[mime] = &mimeMarshaler{inbound: inbound}
-	return nil
-}
+	m.mimeMap[mime] = marshaler
 
-// addOutBound adds an outbund marshaler for a case-sensitive MIME type string ("*" to match any MIME type).
-// Outbound is the marshaler that is used when marshaling outbound responses to the client.
-func (r *marshalerRegistry) addOutbound(mime string, outbound Marshaler) error {
-	mime = http.CanonicalHeaderKey(mime)
-	if mime == "" {
-		return errors.New("empty MIME type")
-	}
-	if entry := (*r)[mime]; entry != nil {
-		entry.outbound = outbound
-		return nil
-	}
-	(*r)[mime] = &mimeMarshaler{outbound: outbound}
 	return nil
-
 }
 
-func (r *marshalerRegistry) lookup(mime string) *mimeMarshaler {
-	if r == nil {
-		return nil
+// MakeMarshalerMIMERegistry returns a new registry of marshalers.
+// It allows for a mapping of case-sensitive Content-Type MIME type string to runtime.Marshaler interfaces.
+//
+// For example, you could allow the client to specify the use of the runtime.JSONPb marshaler
+// with a "applicaton/jsonpb" Content-Type and the use of the runtime.JSONBuiltin marshaler
+// with a "application/json" Content-Type.
+// "*" can be used to match any Content-Type.
+// This can be attached to a ServerMux with the marshaler option.
+func MakeMarshalerMIMERegistry() marshalerRegistry {
```

Comment:
With the new API with `WithMarshalerOption`, you don't need to export this function because the only caller of this function is `NewServeMux`.
Also the function should be named `newMarshalerRegistry` by convention.